File: 2015/01/part_2.rb
# Usage: From root folder
# $ ruby 2015/01/part_1.rb 2015/01/input

DIRS = {"(" => 1, ")" => -1}

def final_floor dirs
  dirs.inject(0){|loc, dir| loc + DIRS[dir]}
end

def reach_basement dirs, pos = 0
  return pos+1 if final_floor(dirs[0..pos]) == -1

  reach_basement(dirs, pos+1)
end

File.open(ARGV[0]).readlines.each do |line|
  puts reach_basement(line.strip.chars)
end

File: 2015/01/part_2.py
#  Usage: From root folder
#  $ python 2015/01/part_2.py 2015/01/input

import sys

DIRS = {"(": 1, ")": -1}


def reach_basement(dirs):
    reach_basement = 0
    for dir in dirs:
        reach_basement += DIRS[dir]
        yield reach_basement
        if reach_basement == -1:
            break


if __name__ == "__main__":
    with open(sys.argv[1]) as file:
        for line in file.readlines():
            print(len(list((reach_basement(line.strip())))))
